bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BDecimal.java

```diff
@@ -94,6 +131,203 @@ public BValue copy(Map<BValue, BValue> refs) {
         return this;
     }
 
+    //========================= Mathematical operations supported ===============================
+
+    public BDecimal add(BDecimal augend) {
+        switch (this.valueKind) {
+            case ZERO:
+                return augend;
+
+            case POSITIVE_INFINITY:
+                if (augend.valueKind == DecimalValueKind.NEGATIVE_INFINITY ||
+                        augend.valueKind == DecimalValueKind.NOT_A_NUMBER) {
+                    return NaN;
+                }
+                return POSITIVE_INF;
+
+            case NEGATIVE_INFINITY:
+                if (augend.valueKind == DecimalValueKind.POSITIVE_INFINITY ||
+                        augend.valueKind == DecimalValueKind.NOT_A_NUMBER) {
+                    return NaN;
+                }
+                return NEGATIVE_INF;
+
+            case NOT_A_NUMBER:
+                return NaN;
+
+            default:
+                if (augend.valueKind == DecimalValueKind.ZERO || augend.valueKind == DecimalValueKind.OTHER) {
+                    return new BDecimal(this.decimalValue().add(augend.decimalValue(), MathContext.DECIMAL128));
+                }
+                return augend;
+        }
+    }
+
+    public BDecimal subtract(BDecimal subtrahend) {
+        switch (this.valueKind) {
+            case ZERO:
+                if (subtrahend.valueKind == DecimalValueKind.ZERO ||
+                        subtrahend.valueKind == DecimalValueKind.NOT_A_NUMBER) {
+                    return subtrahend;
```

Comment:
I think we need always create new BDecimal object with subtrahend. ex. 

    decimal a = 0;
    decimal b = 2.5;
    decimal c = b - a ;
b = 4.5;
we need to check c = 2.5 and b = 4.5. I assume if we not do that b values change will reflect to c. WDYT?

 We need to handle that in other places where we directly return this or operend value.

Reply:
Same decimal object hasn't been modified anywhere. Whenever, we need a different value we simple create a `new BDecimal()` object. Further, I see no reason to modify/change a value of a decimal obejct in future as well. Hence, in the above scenarios, rather than creating a new object returning the same would be more efficient. WDYT?

Reply:
+1